autosklearn/ensemble_building/builder.py

```diff
@@ -0,0 +1,925 @@
+
+            wall_time_in_s = int(time_left - time_buffer)
+            if wall_time_in_s < 1:
+                break
+
+            context = multiprocessing.get_context(pynisher_context)
+            preload_modules(context)
+
+            safe_ensemble_script = pynisher.enforce_limits(
+                wall_time_in_s=wall_time_in_s,
+                mem_in_mb=self.memory_limit,
+                logger=self.logger,
+                context=context,
+            )(self.main)
+
+            safe_ensemble_script(time_left, iteration)
+
+            status = safe_ensemble_script.exit_status
+            if isinstance(status, pynisher.MemorylimitException):
+                # if ensemble script died because of memory error,
+                # reduce nbest to reduce memory consumption and try it again
+
+                # ATTENTION: main will start from scratch;
+                # all data structures are empty again
+                try:
+                    self.previous_candidates_path.unlink()
+                except:  # noqa E722
+                    pass
+
+                if (
+                    isinstance(self.ensemble_nbest, numbers.Integral)
+                    and self.ensemble_nbest <= 1
+                ):
+                    if self.read_at_most == 1:
+                        self.logger.error(
+                            "Memory Exception -- Unable to further reduce the number"
+                            " of ensemble members and can no further limit the number"
+                            " of ensemble members loaded per iteration, please restart"
+                            " Auto-sklearn with a higher value for the argument"
+                            f" `memory_limit` (current limit is {self.memory_limit}MB)."
+                            " The ensemble builder will keep running to delete files"
+                            " from disk in case this was enabled.",
+                        )
+                        self.ensemble_nbest = 0
+                    else:
+                        self.read_at_most = 1
+                        self.logger.warning(
+                            "Memory Exception -- Unable to further reduce the number of"
+                            " ensemble members. Now reducing the number of predictions"
+                            " per call to read at most to 1."
+                        )
+                else:
+                    if isinstance(self.ensemble_nbest, numbers.Integral):
+                        self.ensemble_nbest = max(1, int(self.ensemble_nbest / 2))
+                    else:
+                        self.ensemble_nbest = self.ensemble_nbest / 2
+                    self.logger.warning(
+                        "Memory Exception -- restart with "
+                        "less ensemble_nbest: %d" % self.ensemble_nbest
+                    )
+                    return [], self.ensemble_nbest
+            elif isinstance(status, pynisher.AnythingException):
```

Comment:
When the status is not a `MemoryLimitException`, for example, a `TimeoutException` occurs in the pynished function, a different error will occur.

This then did:
```python
else:
    return safe_ensemble_script.result
```
... but we're not guaranteed that there is a result that looks like `(history, ensemble_nbest)`.

This basically just ensures that if the pynished function raises and can't give back a result then at least we get back the right type. This came up while doing individual tests for the `EnsembleBuilder`.

This was not an issue because `EnsembleBuilderManager` will just see "oh no proper answer from the builder, return something that looks correct, i.e. `([], ensemble_nbest)`. However without the `EnsembleBuilderManager` wrapping it, it was an issue.